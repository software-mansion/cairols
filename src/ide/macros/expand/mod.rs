use std::sync::Arc;

use cairo_lang_defs::db::DefsGroup;
use cairo_lang_defs::plugin::MacroPluginMetadata;
use cairo_lang_filesystem::db::FilesGroup;
use cairo_lang_filesystem::ids::{CrateId, FileId, FileKind, FileLongId, VirtualFile};
use cairo_lang_parser::db::ParserGroup;
use cairo_lang_syntax::node::ast::ModuleItem;
use cairo_lang_syntax::node::kind::SyntaxKind;
use cairo_lang_syntax::node::{SyntaxNode, TypedSyntaxNode};
use cairo_lang_utils::Intern;
use lsp_types::TextDocumentPositionParams;

use crate::lang::db::{AnalysisDatabase, LsSemanticGroup, LsSyntaxGroup};
use crate::lang::lsp::{LsProtoGroup, ToCairo};
use expr::expand_inline_macros_to_file;
use format::format_output;
use inlining::{inline_files, span_after_inlining};
use itertools::Itertools;
use module_item::expand_module_item_macros;

mod expr;
mod format;
mod inlining;
mod module_item;

/// Tries to expand macro, returns it as string.
pub fn expand_macro(db: &AnalysisDatabase, params: &TextDocumentPositionParams) -> Option<String> {
    let file_id = db.file_for_url(&params.text_document.uri)?;
    let node = db.find_syntax_node_at_position(file_id, params.position.to_cairo())?;

    let crate_id = db.find_module_file_containing_node(node)?.0.owning_crate(db);
    let cfg_set = db
        .crate_config(crate_id)
        .and_then(|cfg| cfg.settings.cfg_set.map(Arc::new))
        .unwrap_or(db.cfg_set());
    let edition = db.crate_config(crate_id).map(|cfg| cfg.settings.edition).unwrap_or_default();

    let metadata = MacroPluginMetadata {
        cfg_set: &cfg_set,
        declared_derives: &db.declared_derives(crate_id),
        allowed_features: &Default::default(),
        edition,
    };

    let item_node = node.ancestors_with_self(db).find(|node| {
        let kind = node.kind(db);
        ModuleItem::is_variant(kind) || kind == SyntaxKind::ExprInlineMacro
    })?;

    expand_macro_ex(db, file_id, crate_id, &metadata, item_node)
}

// Recursively expand macros.
fn expand_macro_ex(
    db: &AnalysisDatabase,
    file_to_process: FileId,
    crate_id: CrateId,
    metadata: &MacroPluginMetadata<'_>,
    item_node: SyntaxNode,
) -> Option<String> {
    let mut extra_files = vec![];

    let file_with_inlined_module_items = if item_node.kind(db) == SyntaxKind::ExprInlineMacro {
        file_to_process
    } else {
        // If it is not inline-expr macro, start with expanding regular plugins as these have to be done *before* inline-expr.
        let module_item = ModuleItem::from_syntax_node(db, item_node);

        let mut files = vec![];
        expand_module_item_macros(
            db,
            module_item,
            crate_id,
            metadata,
            &mut files,
            &mut extra_files,
        )?;

        // Inline all files that should be inlined, keep others in `extra_files` for further processing.
        let replaced_content = inline_files(db, file_to_process, &files)?;

        FileLongId::Virtual(VirtualFile {
            parent: None,
            name: "macro_expand".into(),
            content: replaced_content.into(),
            code_mappings: Default::default(),
            kind: FileKind::Module,
            original_item_removed: false,
        })
        .intern(db)
    };

    // Process `extra_files` by collecting all module items in these files and expanding same way as user file.
    let extra_files = extra_files
        .into_iter()
        .filter_map(|content| {
            let expanded_items = db
                .file_module_syntax(content.file)
                .ok()?
                .items(db)
                .elements(db)
                .into_iter()
                .map(|item| {
                    expand_macro_ex(db, content.file, crate_id, metadata, item.as_syntax_node())
                        .unwrap()
                })
                .join("\n\n");

            Some(expanded_items)
        })
        .collect::<Vec<_>>();

    // Expand inline macros only in this span, this prevents expansion of unwanted items
    let span = span_after_inlining(
        db,
        file_to_process,
        file_with_inlined_module_items,
        item_node.span(db),
    )?;

    let replaced_content_file =
        expand_inline_macros_to_file(db, crate_id, file_with_inlined_module_items, span, metadata)?;

    // Span of replaced content in new generated file.
    let new_span =
        span_after_inlining(db, file_to_process, replaced_content_file, item_node.span(db))?;

    let new_file_content = db.file_content(replaced_content_file)?;
    let replaced_content = new_span.take(&new_file_content);

    let replaced_content = if extra_files.is_empty() {
        replaced_content.to_string()
    } else {
        format!("{replaced_content}\n//-----\n{}", extra_files.join("\n//-----\n"))
    };

    Some(format_output(&replaced_content, item_node.kind(db)))
}
