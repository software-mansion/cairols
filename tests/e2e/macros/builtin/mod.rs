use indoc::indoc;

use crate::macros::{fixtures::ProjectWithCairoProjectToml, test_macro_expansion_and_diagnostics};

#[test]
fn no_macro() {
    test_macro_expansion_and_diagnostics!(
        project = ProjectWithCairoProjectToml,
        cwd = "test_package",
        files {
            "test_package/src/lib.cairo" => indoc!(r#"
                fn foo() { <caret> }
            "#)
        }
    );
}

#[test]
fn inline() {
    test_macro_expansion_and_diagnostics!(
        project = ProjectWithCairoProjectToml,
        cwd = "test_package",
        files {
            "test_package/src/lib.cairo" => indoc!(r#"
                fn fib<caret>(mut n: u32) -> u32 {
                   <caret>  pr<caret>intln!<caret>("some text")<caret>;<caret>
                    let mut a: u32 = 0;
                    let mut b: u32 = 1;
                    while n != 0<caret> {
                        n = n - 1;
                        let temp = b;
                        b = a + b;
                        a = t<caret>emp;
                    };
                    a
                }
            "#)
        }
    );
}

#[test]
fn inline_assert_macro() {
    test_macro_expansion_and_diagnostics!(
        project = ProjectWithCairoProjectToml,
        cwd = "test_package",
        files {
            "test_package/src/lib.cairo" => indoc!(r#"
                #[cfg(test)]
                mod tests {
                    #[test]
                    fn test_nothing() {
                        assert<caret>!(1 == 1, "Failure message");
                    }
                }
            "#)
        }
    );
}

#[test]
fn attribute_with_inline() {
    test_macro_expansion_and_diagnostics!(
        project = ProjectWithCairoProjectToml,
        cwd = "test_package",
        files {
            "test_package/src/lib.cairo" => indoc!(r#"
                #[derive(Drop)]
                struct A {
                    a: felt252
                }

                #[genera<caret>te_trait]
                impl ImplA of A<caret>Trait {
                    fn c(self: A) {
                        print<caret>ln!("a");<caret>
                        self.a + 1;
                    }
                }
            "#)
        }
    );
}

#[test]
fn derive() {
    test_macro_expansion_and_diagnostics!(
        project = ProjectWithCairoProjectToml,
        cwd = "test_package",
        files {
            "test_package/src/lib.cairo" => indoc!(r#"
                #[d<caret>erive(Drop,<caret> Serde)]<caret>
                struct <caret>A {
                    a: felt252
                }
            "#)
        }
    );
}
