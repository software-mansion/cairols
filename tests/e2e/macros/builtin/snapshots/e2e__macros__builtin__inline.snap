---
source: tests/e2e/macros/builtin/mod.rs
description: "// → test_package/cairo_project.toml\n[crate_roots]\ntest_package = \"src\"\n\n[config.global]\nedition = \"2024_07\"\n\n// → test_package/src/lib.cairo\nfn fib(mut n: u32) -> u32 {\n     println!(\"some text\");\n    let mut a: u32 = 0;\n    let mut b: u32 = 1;\n    while n != 0 {\n        n = n - 1;\n        let temp = b;\n        b = a + b;\n        a = temp;\n    };\n    a\n}"
expression: report
---
[[expansions]]
analyzed_lines = '''
fn fib<caret>(mut n: u32) -> u32 {
     println!("some text")<caret>;
     println!("some text");<caret>
    while n != 0<caret> {
        a = t<caret>emp;
'''
generated_code = """
// lib.cairo
// ---------

fn fib(mut n: u32) -> u32 {
    {
        let mut __formatter_for_print_macros__: core::fmt::Formatter =
            core::traits::Default::default();
        core::result::ResultTrait::<
            (), core::fmt::Error,
        >::unwrap(
            {
                core::byte_array::ByteArrayTrait::append_word(
                    ref __formatter_for_print_macros__.buffer, 0x736f6d6520746578740a, 10,
                );
                core::result::Result::<(), core::fmt::Error>::Ok(())
            },
        );
        core::debug::print_byte_array_as_string(@__formatter_for_print_macros__.buffer);
    }
    let mut a: u32 = 0;
    let mut b: u32 = 1;
    while n != 0 {
        n = n - 1;
        let temp = b;
        b = a + b;
        a = temp;
    }
    a
}
"""

[[expansions]]
analyzed_lines = '''
   <caret>  println!("some text");
     pr<caret>intln!("some text");
     println!<caret>("some text");
'''
generated_code = '''
// lib.cairo
// ---------

{
    let mut __formatter_for_print_macros__: core::fmt::Formatter = core::traits::Default::default();
    core::result::ResultTrait::<
        (), core::fmt::Error,
    >::unwrap(writeln!(__formatter_for_print_macros__, "some text"));
    core::debug::print_byte_array_as_string(@__formatter_for_print_macros__.buffer);
}
'''
